pkgdb/src/buildenv/realise.cc

```diff
@@ -604,6 +604,49 @@ getRealisedPackages( nix::ref<nix::EvalState> &         state,
 }
 
 
+/* -------------------------------------------------------------------------- */
+
+void
+addScriptToScriptsDir( const std::string &           scriptContents,
+                       const std::filesystem::path & scriptsDir,
+                       const std::string &           scriptName,
+                       std::stringstream & mainActivationScriptContents,
+                       const bool          shouldSource )
+{
+  /* Ensure that the "activate" subdirectory exists */
+  std::filesystem::create_directories( scriptsDir / ACTIVATION_SUBDIR_NAME );
+  std::filesystem::path scriptPath( nix::createTempFile().second );
+  debugLog(
+    nix::fmt( "created tempfile for activation script: script=%s, path=%s",
+              scriptName,
+              scriptPath ) );
+  std::ofstream scriptTmpFile( scriptPath );
+  scriptTmpFile << scriptContents;
+  scriptTmpFile.close();
+  std::filesystem::copy_file( scriptPath,
+                              scriptsDir / ACTIVATION_SUBDIR_NAME
+                                / scriptName );
+  std::filesystem::permissions( scriptsDir / ACTIVATION_SUBDIR_NAME
+                                  / scriptName,
+                                std::filesystem::perms::owner_exec,
+                                std::filesystem::perm_options::add );
+  auto scriptRelativePath
+    = nix::fmt( "\"$FLOX_ENV/%s/%s\"", ACTIVATION_SUBDIR_NAME, scriptName );
+  if ( shouldSource )
+    {
+      mainActivationScriptContents << "source " << scriptRelativePath << '\n';
+    }
+  else
+    {
+      mainActivationScriptContents << "bash " << scriptRelativePath << '\n';
+    }
```

Comment:
Should this be rather two separate operations?
- add a script to the closure
- construct the main activate script

passing a stringstream into this method _and_ conditionally add to that stream is a bit weird IMO

Reply:
I'd prefer that adding the script to the directory and sourcing it was a single function call so you're never in a state in which one is filled in and the other isn't.